Feature/cppcommon

tarc opened the pull request:
Specify library name and version:  **cppcommon/1.0.0.0**

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/conan-io/conan-center-index?pullRequest=3224) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/conan-io/conan-center-index?pullRequest=3224) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: tarc<br/>:x: Tarcisio Rodrigues<br/><hr/>**Tarcisio Rodrigues** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/conan-io/conan-center-index?pullRequest=3224) it.</sub>

prince-chrismc commented on recipes/cppcommon/all/conandata.yml:
```diff
@@ -0,0 +1,8 @@
+sources:
+  "1.0.0.0":
+    url: "https://github.com/tarc/CppCommon/archive/d75b3447ecc89289539740a05fed618fa9cae4f6.tar.gz"
```
```suggestion
    url: "https://github.com/tarc/CppCommon/archive/1.0.0.0.tar.gz"
```

tarc replied:
fixed by 9a3f4fd

prince-chrismc commented on recipes/cppcommon/all/conanfile.py:
```diff
@@ -0,0 +1,97 @@
+
+    def source(self):
+        tools.get(**self.conan_data["sources"][self.version])
+        os.rename("CppCommon-" + "d75b3447ecc89289539740a05fed618fa9cae4f6",
```
```suggestion
        os.rename("CppCommon-" + self.version,
```

tarc replied:
Fixed by 9a3f4fd

prince-chrismc commented on recipes/cppcommon/all/conanfile.py:
```diff
@@ -0,0 +1,97 @@
+            self.output.warn("cppcommon requires C++17. Your compiler is unknown. Assuming it supports C++17.")
+
+    def build(self):
+        patches = self.conan_data["patches"][self.version]
+        for patch in patches or []:
+            tools.patch(**patch)
+
```
```suggestion
        for patch in self.conan_data.get("patches", {}).get(self.version, []):
            tools.patch(**patch)
```

prince-chrismc commented on recipes/cppcommon/all/conanfile.py:
```diff
@@ -0,0 +1,97 @@
+
+    def package(self):
+        cmake = self._configure_cmake()
+        cmake.install()
+        include = os.path.join(self._source_subfolder, "include")
+        self.copy("*.h", src=include, dst="include", keep_path=True)
+        self.copy("*.inl", src=include, dst="include", keep_path=True)
```
I did not check upstream yet, but it seems odd to cmake install and copy headers :thought_balloon:

tarc replied:
Fixed in 809da42

prince-chrismc commented on recipes/cppcommon/all/test_package/CMakeLists.txt:
```diff
@@ -0,0 +1,11 @@
+cmake_minimum_required(VERSION 3.1)
```
CXX 17 was added in 3.8
```suggestion
cmake_minimum_required(VERSION 3.8)
```

prince-chrismc reviewed (changes requested):
You are using your fork, is there a reason not to use https://github.com/chronoxor/CppCommon ?

tarc commented:
> You are using your fork, is there a reason not to use https://github.com/chronoxor/CppCommon ?

Sorry, I got carried away while trying to get the library to work properly on all platforms. I changed to my fork for I was checking it with github pipelines before pushing this PR - I had to fix a lot. I intend now to submit a PR to the main repo.

I'm addressing some of your comments and marking this as a draft. I think it's better to wait for the main author to answer to the PRs on the main repo before proceeding.

tarc commented:
So, @prince-chrismc , I moved back to the main repo and added a huge patch file for the changes that were on my fork.

Now I'm submitting a PR to chronoxor/CppCommon with those changes. If it got through all is left to do is:

- see to the copying of header files you pointed
- update version in `conandata.yml` and `config.yml`
- remove the second patch

prince-chrismc commented:
https://github.com/chronoxor/CppCommon#features says cross platform first :hatching_chick: when you open your PR against upstream, mention this one to help highlight the issues =)

tarc commented:
Hey, @prince-chrismc . I got ir wrong. Actually the code needs verry little to be fixed to work on all platforms being built by the pipeline.

If you consider the patch directory it has now only some minor fixes.

I've already opened two PRs upstream, but maybe we should go ahead with this one and maybe add another version afterwards, when the author fixes thes issues.

prince-chrismc commented:
> I've already opened two PRs upstream, but maybe we should go ahead with this one and maybe add another version afterwards, when the author fixes thes issues.

That's a very good option, I've seen it done here in the past, go for it!

prince-chrismc commented on recipes/cppcommon/all/patches/00001-update-cmakelists.patch:
```diff
@@ -0,0 +1,66 @@
+-# External packages
+-find_package(Threads REQUIRED)
+-if(UNIX AND NOT APPLE AND NOT MSYS)
+-  find_package(LibBFD)
+-  find_package(LibDL)
+-  find_package(LibRT)
+-  find_package(LibUUID)
+-endif()
+-if(WIN32 OR MSYS)
+-  find_package(DbgHelp)
+-  find_package(RPC)
+-  find_package(Userenv)
```
I see some of the system deps are listed in the recipe, can you confirm the rest are not required? ie its not linking against the build image ones

tarc replied:
Estava fazendo alguns testes com os exemplos e benchmarks da biblioteca pra validar isso...

uilianries commented on recipes/cppcommon/all/conandata.yml:
```diff
@@ -0,0 +1,17 @@
+  "1.0.0.0":
+    url: "https://github.com/chronoxor/CppCommon/archive/1.0.0.0.tar.gz"
+    sha256: "d6124dd4cd430e5f10c2942ff02a95636369875b652e66a499b01f1fb3ecfd6d"
+  "19265a8d721e9c8d5f706c11afbd8885fa9ee9e2":
```
The release 1.0.0 was 18 days ago, I don't think adding a custom version now is a good idea. Ask for a new release if you really need a new fix/feature.

tarc replied:
This project actually doesn't have an usual release cycle. The author keeps commiting in the master branch and the pipelines check the whole thing. This 1.0.0.0 version was added after I opened an issue (https://github.com/chronoxor/CppCommon/issues/4).

I'm thinking that maybe we should stick to a different versioning scheme here on conan center that is more suited to the way the project handles its own versioning.

In [another PR of mine](#3380), @SpaceIm suggested something I didn't know: to use a version number like this `cci.20171104` to represent _the version_ that was up on the date 2017 11 04.

For instance, to get the 1.0.0.0 version of this repo working properly with all the pipelines covered here on conan center, it needs some patching. I think it would be better to skip this version altogether and go along with the way the author updates upstream, selecting some specific _checkpoints_ from which a version is generated.

In this case, the version `cli.20201029` (which is 7 commits ahead of 1.0.0.0) would need only a patch on the `CMakeLists.txt` to integrate with Conan. Although the 1.0.0.0 version needs some non trivial bug fixing.

As the author itself doesn't seem to care to some traditional versioning scheme, isn't it better to use the `cli.20201029` approach instead?

uilianries replied:
Please, read: https://github.com/conan-io/conan-center-index/blob/master/docs/faqs.md#what-version-should-packages-use-for-libraries-without-official-releases

If cppcommon offers an official release, it should be added in conandata.yml, only if it's broken you could avoid, but otherwise it should be added.

And, either the upstream doesn't offer regular releases, because prefers bleeding-edge versioning, use the recommended version which is documented on FAQs.

tarc replied:
Fixed in 2e6b9154, I think...

prince-chrismc commented on recipes/cppcommon/all/CMakeLists.txt:
```diff
@@ -0,0 +1,38 @@
+cmake_minimum_required(VERSION 3.13)
+project(cmake_wrapper)
+
+include(${CMAKE_CURRENT_BINARY_DIR}/../conanbuildinfo.cmake)
```
```suggestion
include(conanbuildinfo.cmake)
```

tarc replied:
> Very close!
> 
> 🙏 please refrain from squashing it make's it harder for us to see the changes

Sorry! I didn't know!

prince-chrismc commented on recipes/cppcommon/all/CMakeLists.txt:
```diff
@@ -0,0 +1,38 @@
+cmake_minimum_required(VERSION 3.13)
```
I think this could be lower to 3.8 since it's C++17, I dont see anything that requires higher

tarc replied:
Lowering it to 3.8 gives this cmake error:

> CMake Error at CMakeLists.txt:14 (target_link_libraries):
  Attempt to add link library "fmt" to target "cppcommon" which is not built
  in this directory.
>
>  This is allowed only when policy CMP0079 is set to NEW.

The line in question is:

https://github.com/conan-io/conan-center-index/blob/b87568ce3922fd7c99eb3a9ae1958049c9b3f031/recipes/cppcommon/all/CMakeLists.txt#L14

Complaining that the wrapper is linking the dependencies to the upstream target...

prince-chrismc reviewed (commented):
Very close!

🙏 please refrain from squashing it make's it harder for us to see the changes

prince-chrismc reviewed (changes requested):
Please consider https://github.com/tarc/conan-center-index/pull/3

prince-chrismc commented on recipes/cppcommon/all/patches/00001-update-cmakelists.patch:
```diff
@@ -0,0 +1,128 @@
+   find_package(LibDL)
+   find_package(LibRT)
+-  find_package(LibUUID)
++  find_package(libuuid)
```
Can you please put a fix me here so we know this is c3i issue

```suggestion
+  find_package(libuuid) #FIXME: CCI does not generate correctly #3441
```

prince-chrismc reviewed (commented):
Sorry for the trouble, I had it installed locally so I didnt have any issues 🙈

Super close, just add that comment and I can approve!

prince-chrismc reviewed (approved):
:rocket:
